Add basic support for element selection

adsamcik opened the pull request:
Adds very basic support for element selection that has to be manually entered in page settings so it will mainly be for advanced users, however it should be better than nothing. Supports #id_name for id selection, .class_name for class selection, [index] for picking specific element from matched id's and nesting so that #id_name.class_name[index] is also possible. Multiple conditions can also be entered by separating them with comma.

There are possible options for the future like whether change in number of elements should be considered a change or not. But this should be somewhat good start.

I wasn't able to test it properly because of issues with firefox developer edition see #595

adsamcik commented:
I found some more time so I added contentMode which allows more control over how changes are detected but I had some trouble with the tests. 

I moved some parts of what scan did to new scan_content and moved appropriate tests, but it seems these tests are not called by the karma or whatever. I have no experience with this engine so I left it as is for now, because I can't fix the tests if they are not called at all even though I couldn't find a reason why. The only clue is some 404 errors, but I didn't find out why they are there either.

I also fixed some documentation issues with types and stuff.

sneakypete81 commented:
> I wasn't able to test it properly because of issues with firefox developer edition see #595
...
> I moved some parts of what scan did to new scan_content and moved appropriate tests, but it seems these tests are not called by the karma or whatever.

I'll try to find time to look at this - unfortunately I'm pretty busy at the moment.

Thanks again for all of your efforts with this, I really appreciate it.

adsamcik commented:
> I think fundamentally the problem this is trying to solve is: "I'd like to scan just part of a page, without having changes in other parts cause a notification". Is that a reasonable goal for this?

Yes pretty much. 

Are you sure about the MINOR_CHANGE? It would require full page scan to run after the part scans if no change is detected.

Point and click UI was something I considered but in the end it seemed like way too time consuming. One of the things I really wanted was to match multiple elements which would be even more difficult with UI. That's why I called it "basic" in the end.

The approach you mentioned is basically what this does.

I expected ContentMode, MatchMode, MatchCount to be considered out of scope, but if UI is the main problem, wouldn't something like advanced toggle solve it? 

I added match count and content mode because those are the features I could use for my use case. Because what I want is to just check if the number of matched elements hasn't changed. For that I can choose match count true and content mode ignore content (not implemented yet). The reason is that inside the content there can be changes, but I don't care about them. MatchMode is more like an experiment that I personally don't really need.

I would like to note that I don't plan on adding anything else but I really would like to keep MatchCount and ContentMode.

adsamcik commented:
I figured out why the tests weren't working. It was caused by auto-imports ignoring .js at the end of each import. So all the tests should now be passing (number of tests seem to be the same to your current version).

I also noticed that numbers do not exactly strip just numbers (which is somewhat unexpected behaviour) so I changed that as well so it should now keep [,.] that are probably not number separators (100,000 will get stripped as a whole ofc).

sneakypete81 commented:
Makes sense, thanks for clarifying this.

Taking one extreme view: fewer features and less code can help prevent bugs, reduce maintenance, make future changes easier, reduce user confusion, etc.etc. But there's some great suff here, I appreciate the effort you're putting in and would really like to merge this!

How about something like a combobox with the following options:

Trigger on changes:
 * Anywhere in the page (default)
 * Inside the following HTML elements:
 * That change the number of HTML elements (content is ignored):

Then below that a text input labelled "HTML element selector" (including some explanatory "usage" description like you've already done) that is disabled if the first combo item is selected.

Does that cover everything? I think this would make this feature easier to understand and use.

sneakypete81 commented:
A future enhancement might be to add another option to the combobox to exclude certain areas from change detection:
 * Outside the following HTML elements:

sneakypete81 commented:
> Are you sure about the MINOR_CHANGE? It would require full page scan to run after the part scans if no change is detected.

MINOR_CHANGE is really just to differentiate between no change and a change that doesn't trigger a notification. Other parts of the code then use this to decide whether to update the stored HTML. That's why it's important that the semantics are maintained.

An extra scan shouldn't be needed. From the original `scan.getChangeType` function:

```javascript
  if (str1 == null) {
    // This is the first scan.
    return changeEnum.NEW_CONTENT;
  } else if (str1 === str2) {
    // HTML is unchanged.
    return changeEnum.NO_CHANGE;

  } else if (__.isMajorChange(str1, str2, changeThreshold)) {
    // Change is larger than changeThreshold.
    return changeEnum.MAJOR_CHANGE;

  } else {
    // Change is smaller than changeThreshold.
    return changeEnum.MINOR_CHANGE;
  }
```

The first two clauses are trivial - exit early if there's nothing to do.

To add your functionality, just modify the third "isMajorChange" clause with a function call that does one of three things:
 1. In "traditional" mode, return the result of `fuzzy.isMajorChange` (as with the code above).
 2. In "just inside these elements" mode, remove all content outside the chosen elements, then return the result of `fuzzy.isMajorChange`.
 3. In "element count" mode, check the element count and return true if it has changed.

The final clause will return MINOR_CHANGE for a small change that didn't trigger a change notification.

sneakypete81 commented:
> I figured out why the tests weren't working. It was caused by auto-imports ignoring .js at the end of each import.

Well spotted!

> I also replaced the karma-notification-reporter with karma-notify-reporter

Thanks, I don't develop much on Windows anymore, so this is a good move. In future this kind of thing should probably have its own PR if possible, since small self-contained PRs are a bit easier to review and merge - don't worry about it this time though :-).

adsamcik commented:
> Taking one extreme view: fewer features and less code can help prevent bugs, reduce maintenance, make future changes easier, reduce user confusion, etc.etc.

Well in this case I wouldn't say it's more code and I think it's easier to understand like this. What I agree with is that currently the UI is not that easy to understand. I think the dropdown (combo box) would indeed make things clearer, but I think the options should still be preserved under some advanced mode, but I will probably leave that for the future. So tomorrow if I find the time I'll collapse everything into single dropdown, not gonna touch the logic, it's fine as is imo. 

The logic should be basically done if it's working as it should be so the UI is really the only thing left.

> In future this kind of thing should probably have its own PR if possible, since small self-contained PRs are a bit easier to review and merge - don't worry about it this time though :-).

I agree.

> An extra scan shouldn't be needed. From the original scan.getChangeType function:

At this moment, the getChangeType function is no more. I replaced it with getChanges which already does everything it is supposed to. The removal was rather an evolution rather than decision I simply realized over time it was no longer even used.

adsamcik commented:
I added tests and updated the UI.

I also found one more thing for the future and that is option to cumulate change such that if you match 3 elements with id example and each by itself is minor change, but if added together they become major change. It could be done by merging the matches into a single string, but I am not sure it will work as it should.

Also nice option for the future might be highlighting only changes within the change matched html.

Also during my testing if UI works I found that when single page is scanned due to some delay somewhere it is not reflected in a notification. This seems to be related to storage (which I saw you are slowly reworking in spare time). The issue seems to be caused by some delay with what this line does this.pageStore.getChangedPageList().length. but since I noticed it I thought I'd mention it.

adsamcik commented:
After writing tests for the matcher I realized it was going nowhere like this so in the end I ended up scraping it and replacing it with dom construction and it's query. This makes it much simpler and also greatly increases it's capabilities. The rest works the same as it did before.

sneakypete81 commented:
Let me know when this is ready for a review, thanks for all of your efforts with this and your other PRs.

adsamcik commented:
It should be. I've been using it for awhile and haven't found any problems past some I already fixed.

sneakypete81 commented:
Thanks for the formatting and equality fixes. I was struggling a bit to see your "real" changes intertwined with these, so have merged these obvious improvements first.

Could you please rebase the remainder of this PR onto master?

adsamcik commented:
Done I hope.

sneakypete81 commented on src/lib/scan/selector_matcher.js:
```diff
@@ -0,0 +1,24 @@
+  const result = [];
+  matches.forEach((element) => result.push(element.outerHTML));
+  return result;
+}
```
(not something for now, maybe later)

adsamcik replied:
That's exactly what I thought, but I considered it out of scope so I ignored it.

adsamcik commented on src/app/main/main.css:
```diff
@@ -169,6 +175,10 @@ input {
+select {
+  width: 100%;
+}
+
```
Yeah I forgot to update the wording after adding the description. The width however served different purpose that is no longer needed. I'll remove it.

adsamcik replied:
Replaced with max-width and shortened the text

adsamcik commented on src/lib/main/dialog_view.js:
```diff
@@ -150,6 +169,158 @@ function updateAutoscanDescription(sliderValue) {
+<a href="https://www.w3schools.com/cssref/css_selectors.asp">
+https://www.w3schools.com/cssref/css_selectors.asp</a>`;
+  } else {
+    return `Selectors not available in "Anywhere" scan mode.`;
```
Yes, but it might not be clear why.

sneakypete81 replied:
I think it is pretty clear if the Scan Mode input is shifted to just above the CSS selector field (see my mockup above)

adsamcik replied:
Well grouping it under scan mode is really weird and does not make much sense imo since they are totally different things, it just so happens that some scan modes enable/support them. And it's really not clear, I myself wouldn't be sure why it's greyed out, is it cause of a bug? How do I enable it?

I decided to make a compromise so that there is less code around it which seemed like your main goal. It will no longer change the description and instead just set the placeholder, so that new users still immediately know why the feature is disabled and can use the description to look up what it is.

adsamcik commented on src/app/main/main.css:
```diff
@@ -146,6 +146,12 @@ legend {
+  color: rgba(0, 0, 0, 0.8);
+  text-align: left;
+}
+
```
leftover from when it used different tag. Removed.

adsamcik commented on src/app/main/main.css:
```diff
@@ -201,3 +211,7 @@ form button.browser-style {
+
+.disabled {
+  opacity: 0.5;
+}
```
I add disabled to fields that are not supposed to be used. But the sliders did not indicate it in any way that they were disabled.

sneakypete81 replied:
That's all good - I was talking about the input field for the CSS selector. There's no visual indication when it's disabled, so changing the input box's background to light grey might help with this.

adsamcik replied:
hmm, but there is, it is marked as disabled and thus browser greys it out by itself. https://www.w3schools.com/tags/att_input_disabled.asp

sneakypete81 replied:
Maybe it's a MacOS thing, but the disabled control looks the same to me. But the text you've added is greyed out, so it's now very clear. Thanks!

adsamcik replied:
I see, I found https://github.com/uswds/uswds/issues/1856 so it seems that MacOS has different disabled styles for some reason, if the placeholder was enough, that's good.

adsamcik commented on src/app/main/main.html:
```diff
@@ -60,6 +60,19 @@ <h1 id="folder-heading">Folder Settings</h1>
+          </select>
+          <output name="scan-mode-description"></output>
+        </fieldset>
 
```
To be honest, the whole UI kinda needs an overhaul. The labels are very inconsistent, some are to the left some above. Scan mode is right below URL because it affects all the fields below it.

sneakypete81 replied:
Agreed! I'd still prefer the Scan Mode stuff to be all together at the bottom, even if the count mode causes the Change threshold to be disabled further up the page. The goal is to make it as clear as possible for people who don't understand or want to use these new settings.

adsamcik commented on src/lib/scan/scan_content.js:
```diff
@@ -0,0 +1,309 @@
+
+  if (prevData.html === scannedData.html) {
+    return changeEnum.NO_CHANGE;
+  }
```
Comparing strings is much slower operation compared to at most 2 null checks and number compare.

adsamcik commented on src/lib/scan/scan_content.js:
```diff
@@ -0,0 +1,309 @@
+
+  // If no change was detected in parts, just return minor change because we
+  // already know there was a change somewhere in the html
+  if (htmlChange === changeEnum.NO_CHANGE) {
```
This is the check you wanted if any change is detected outside of the selected html elements than report it as minor change. And it needs to strip the tags so that it only compares text.

sneakypete81 replied:
Of course - thanks!

adsamcik commented on src/lib/scan/scan_content.js:
```diff
@@ -0,0 +1,309 @@
+        };
+      }
+    };
+  } else if (matchMode === matchModeEnum.LAST) {
```
Ok.

adsamcik commented on src/lib/main/dialog_view.js:
```diff
@@ -150,6 +169,158 @@ function updateAutoscanDescription(sliderValue) {
+      matchMode: Page.matchModeEnum.FIRST,
+    },
+  }],
+]);
```
There are no invalid combinations of attributes! Any combination outside of the allowed values is still valid (unless it contains wrong type or something like that).

adsamcik replied:
I spent quite a bit of time designing and thinking about this and I must say I am very strongly against the way you propose. The way I designed it is that the options are actually only front end thing for the user. It might add a bit of perceived complexity to the page attributes and an extra resolve function to the UI but it makes for a better code for scanning functions. Also it's easier to extend in the future.

If you had scan mode like you propose you would either need to have duplicate code or complex conditions. Neither of which is particularly appealing.

sneakypete81 replied:
Ok, that's fair enough. Thanks for the explanation.

adsamcik commented on src/lib/main/dialog_view.js:
```diff
@@ -150,6 +169,158 @@ function updateAutoscanDescription(sliderValue) {
+ *
+ * @param {string} modeName - Scan mode name.
+ */
+function updateScanModeDescription(modeName) {
```
I think you got confused. This also updates description for the modes.

sneakypete81 replied:
Ah, right you are. Maybe rename the function to something more descriptive if it does more than update just the scan mode description.

sneakypete81 commented:
Merged in `ee2ee32`.

Many thanks for your persistence with this, I'll get a beta out soon for further testing.
